DOC: add Comparison with Excel

afeld opened the pull request:
- [x] closes https://github.com/pandas-dev/pandas/issues/22993
- [x] tests added / passed
- [ ] ~~passes `black pandas`~~
- [ ] ~~passes `git diff upstream/master -u -- "*.py" | flake8 --diff`~~
- [ ] ~~whatsnew entry~~

## Background

I teach [a class on pandas for public policy students](https://github.com/afeld/python-public-policy/blob/master/syllabus.md#readme), and for many of them, spreadsheets are the only point of reference they have for working with tabular data. It would be very helpful to have (official) document comparing the two to point them to.

This is my first contribution to pandas and first time using reStructuredText, so feedback welcome. Thanks in advance!

## TODOs

Making a running checklist to show what I've done already, and what else I plan to do. Hoping for some preliminary feedback (like is there still interest in having this page) before spending too much more time on it.

Happy to continue in this pull request until complete with all of them, or get this merged sooner than later and take care of the others in follow-up pull requests. Slight preference for the latter (some documentation being better than none, less to review at once, etc.), but open to whatever.

- [x] start with what @rotuna did in https://github.com/pandas-dev/pandas/pull/23042
- [x] add links to canonical Excel documentation, as [previously requested](https://github.com/pandas-dev/pandas/pull/23042#discussion_r224289676)
- [x] fix for CI
- [x] link from Getting Started pages
- [ ] incorporate structure from [SAS](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sas.html)/[STATA](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_stata.html) comparison pages
   - [x] Data structures
   - [ ] Data input / output
      - [x] Mention `read_excel()`
   - [ ] Data operations
   - [ ] String processing
   - [ ] Merging
   - [ ] Missing data
   - [ ] GroupBy
   - [ ] Other considerations
- [ ] convert examples to use [`tips` dataset](https://raw.github.com/pandas-dev/pandas/master/pandas/tests/io/data/csv/tips.csv)
- [ ] add info about charting

## Questions

- [x] Which whatsnew file should I add to?
- [x] I noticed that [`doc/source/_static` is in the `.gitignore`](https://github.com/pandas-dev/pandas/blob/54682234e3a3e89e246313bf8f9a53f98b199e7b/.gitignore#L113), but [there are files checked into that folder](https://github.com/pandas-dev/pandas/tree/master/doc/source/_static). Is that intentional? - https://github.com/pandas-dev/pandas/pull/38739
- [ ] Some of the comparison documentation refers to "columns", while other refer to "Series". Is there a preference, or can they be used interchangeably?
- [x] Since spreadsheet software is largely interchangeable/compatible, would it make sense to make the page more general as "Comparison to spreadsheets"?
- [ ] Thoughts about including [slightly more subjective content](https://colab.research.google.com/github/afeld/python-public-policy/blob/main/pandas_crash_course.ipynb#scrollTo=efp6dYYtL-Je), such as _why_ one might want to use spreadsheets vs. pandas?

rotuna committed:
DOC: Added example for Excel Filters and Fill handle

afeld committed:
DOC: move the comparison with Excel into the new folder

afeld committed:
DOC: add introduction to Excel page

afeld committed:
DOC: have excel page boilerplate match other comparison pages

afeld committed:
DOC: make shared include for the boilerplate text in comparison docs

afeld committed:
DOC: add links to Excel page from Getting Started

afeld committed:
DOC: add section on Data Structures to Excel page

afeld committed:
DOC: fix CI errors on Excel page

- Format Excel comparison code samples with [blacken-docs](https://github.com/asottile/blacken-docs)
- Fix `SettingWithCopyWarning`s

MarcoGorelli commented on doc/source/getting_started/comparison/comparison_boilerplate.rst:
```diff
@@ -0,0 +1,9 @@
+If you're new to pandas, you might want to first read through :ref:`10 Minutes to pandas<10min>`
+to familiarize yourself with the library.
+
+As is customary, we import pandas and NumPy as follows:
+
+.. ipython:: python
+
+    import pandas as pd
+    import numpy as np
```
nice!

MarcoGorelli reviewed (commented):
Thanks @afeld - I built this locally and it generally looks good

To answer some questions:

> Which whatsnew file should I add to?

It shouldn't be necessary to add one, that's usually for new features / bug and regression fixes

> Since spreadsheet software is largely interchangeable/compatible, would it make sense to make the page more general as "Comparison to spreadsheets"?

I think so, perhaps it could be "Comparison with spreadsheets (e.g. Excel)"?

EDIT
----

I don't think this needs a whatsnew note

jreback commented:
@afeld can you post a rendered picture here of the new docs page here

jreback commented on doc/source/getting_started/comparison/comparison_with_excel.rst:
```diff
@@ -0,0 +1,213 @@
+
+    ``DataFrame``, worksheet
+    ``Series``, column
+    ``Index``, row headings
```
should also indicate that the row labels themselves are akin to the default RangeIndex

afeld replied:
Added a mention of `RangeIndex` below. That work?

jreback commented on doc/source/getting_started/comparison/comparison_with_excel.rst:
```diff
@@ -0,0 +1,213 @@
+General terminology translation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. csv-table::
```
it might be useful to include .png's here if it helps explain material

afeld replied:
Sure, will put something together. Ok as a follow-up?

jreback replied:
sure

jreback commented on doc/source/getting_started/comparison/comparison_with_excel.rst:
```diff
@@ -0,0 +1,213 @@
+Drop Duplicates
+~~~~~~~~~~~~~~~
+
+Another commonly used function is Drop Duplicates. This is directly supported in
```
might be good to give references to the pandas functions themselves (and the excel functions if you can link to them)

afeld replied:
Done.

jreback commented on doc/source/getting_started/comparison/comparison_with_excel.rst:
```diff
@@ -0,0 +1,213 @@
+~~~~~~~~~~~
+
+This can be achieved by using ``pandas.pivot_table`` for examples and reference,
+please see `pandas.pivot_table <http://pandas.pydata.org/pandas-docs/stable/generated/pandas.pivot_table.html>`__
```
maybe a mini-example?

afeld replied:
Done.

jreback commented on doc/source/getting_started/comparison/comparison_with_excel.rst:
```diff
@@ -0,0 +1,213 @@
+        return row["student_count"] - row["girls_count"]
+
+
+    df["boys_count"] = df.apply(get_count, axis=1)
```
better to just show this a direct operation (or offer the .apply as an alternative, not the primary), e.g.

```
df['boys_count'] = df['student_count'] - df['girls_counts']
```

afeld replied:
Done.

jreback reviewed (changes requested):
@afeld looks great. ideally if you can add the excel (and pandas) references. am happy to merge (and can keep adding things in later PRs).

afeld committed:
DOC: improve formula documentation under Getting Started

- Mention apply() in documentation around deriving columns
- Simplify code for doing column subtraction in Excel doc

afeld committed:
DOC: link to duplication about dropping duplicates (in Excel and pandas)

afeld committed:
DOC: add examples for importing and pivoting data in Excel

afeld committed:
DOC: make the comparison to Excel about spreadsheets more generally

afeld committed:
DOC: mention RangeIndexes on the spreadsheets page

afeld committed:
DOC: fix long line in spreadsheets comparison

afeld committed:
DOC: fix reference to tips DataFrame on spreadsheets page

afeld commented:
<details>
<summary>Screenshot of the new page</summary>
<img src="https://user-images.githubusercontent.com/86842/103143847-e3065000-46ec-11eb-94d3-3cfdf8579ca5.png" alt="screenshot of comparison to Excel page">
</details>

afeld commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,247 @@
+~~~~~~~~~
+
+Every ``DataFrame`` and ``Series`` has an ``Index``, which are labels on the *rows* of the data. In
+pandas, if no index is specified, a :class:`~pandas.RangeIndex` is used by default (first row = 0,
```
I wasn't quite sure whether to use the `~` in the class/method/function references, so let me know if you want that changed.

jreback replied:
this is fine

afeld commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,247 @@
+
+   url = (
+       "https://raw.github.com/pandas-dev"
+       "/pandas/master/pandas/tests/io/data/csv/tips.csv"
```
I copied this code from another file and kept it consistent, but it wasn't clear why the URL was being passed in as a tuple rather than a single string.

MarcoGorelli replied:
this isn't a tuple, it's still a single string
```python
>>> url = (
...     'a'
...     'b'
... )
>>> url
'ab'
```

afeld replied:
Ah, I missed that there wasn't a comma! Wasn't aware of [implicit string concatenation](https://www.python.org/dev/peps/pep-3126/).

afeld committed:
DOC: fix build warnings on spreadsheet comparison page

MarcoGorelli commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,247 @@
+
+.. ipython:: python
+
+   pd.pivot_table(tips, values="tip", index=["size"], columns=["sex"],
+                  aggfunc=np.average)
```
Nitpick, but just to be consistent with the rest of the docs:
```suggestion
    pd.pivot_table(
        tips, values="tip", index=["size"], columns=["sex"], aggfunc=np.average
    )
```

MarcoGorelli commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,247 @@
+
+    import random
+
+    df1 = pd.DataFrame(
+        {
+            "keys": [1, 2, 3, 4, 5, 6, 7],
+            "first_names": [
+                "harry",
+                "ron",
+                "hermione",
+                "rubius",
+                "albus",
+                "severus",
+                "luna",
+            ],
+        }
+    )
+    df1
```
I think this looks better if you define `first_names` outside
```suggestion
    first_names = [
        "harry",
        "ron",
        "hermione",
        "rubius",
        "albus",
        "severus",
        "luna",
    ]
    df1 = pd.DataFrame(
        {"keys": [1, 2, 3, 4, 5, 6, 7], "first_names": first_names}
    )
    df1
```

MarcoGorelli commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,247 @@
+    )
+    df1
+
+    random_names = pd.DataFrame(
+        {
+            "surnames": [
+                "hadrid",
+                "malfoy",
+                "lovegood",
+                "dumbledore",
+                "grindelwald",
+                "granger",
+                "weasly",
+                "riddle",
+                "longbottom",
+                "snape",
+            ],
+            "keys": [random.randint(1, 7) for x in range(0, 10)],
+        }
+    )
```
similarly here
```suggestion
    surnames = [
        "hadrid",
        "malfoy",
        "lovegood",
        "dumbledore",
        "grindelwald",
        "granger",
        "weasly",
        "riddle",
        "longbottom",
        "snape",
    ]
    keys = [random.randint(1, 7) for x in range(0, 10)]
    random_names = pd.DataFrame({"surnames": surnames, "keys": keys})
```

afeld committed:
DOC: tweaks to code formatting on spreadsheets page

Co-authored-by: Marco Gorelli <m.e.gorelli@gmail.com>

jreback commented on doc/source/getting_started/comparison/comparison_with_spreadsheets.rst:
```diff
@@ -0,0 +1,242 @@
+Importing data
+~~~~~~~~~~~~~~
+
+Both `Excel <https://support.microsoft.com/en-us/office/import-data-from-external-data-sources-power-query-be4330b3-5356-486c-a168-b68e9e616f5a>`__
```
can you also show using read_excel here (

afeld replied:
Added. [The documentation around Excel on the IO page](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-excel) is pretty clear, so linked rather than including an example here.

afeld replied:
Let me know if you feel strongly about having an example in this page directly.

jreback replied:
i think if you are going to do a followup to this PR for the other changes would be useful to have taking that .csv file, reading it in and writing it out using .to_excel then reading it back in with read_excell, e.g. just to give a flavor of things. (but for later).

afeld committed:
DOC: add information about reading/writing Excel files to comparison page

jreback commented:
thanks @afeld very nice. built docs will appear here: https://pandas.pydata.org/pandas-docs/dev/ (may take a little for the CI to do it). happy to have a followup as discussed.

jreback commented:
minor point: https://pandas.pydata.org/pandas-docs/dev/getting_started/comparison/comparison_with_spreadsheets.html#pivot-tables

you can add `margins=True` to have the exact excel pivot table behavior you are showing in the png.
